### ISSUE TITLE
Endless ec2 instance deployments on failed registration

### ISSUE THREAD

**Issue by brettcave** [iriusrisk/GoCD-EC2-Elastic-Agent-Plugin#14]
If an Elastic Agent Profile contains configuration that results in the go-agent not registering successully with Go Server, then an ec2 instance will be provisioned, and not get terminated. A new instance will then be launched to try replace it. The new instance won't register either, and so new instances continuously created. This continues, even if the pipeline run that triggered the EA provisioning is cancelled.

Better behaviour would be to terminate ec2 instances that are launched but don't successfully register within time period.

In short: this can max out AWS instance capacity to limits (service limits, subnet/iP limits, etc).

**Comment by v-sky** [comment]
I'm concerned that your proposed behavior could accidentally terminate not desired instances.
I remember checking thins behavior with the 18 server version, and it stopped creating new instances after the pipe was canceled. I need to check it again with the latest version.

**Comment by brettcave** [comment]
I am running this on 19.10 @v-sky - might be something that works in newer versions.

**Comment by brettcave** [comment]
@v-sky - the sort of logic i was thinking about for terminating agents would be based on setting an expected registration window. The instance ID is expected to be sent to GoCD to provide correlation - `agent.auto.register.elasticAgent.agentId=$(ec2-metadata --instance-id | cut -d " " -f 2)`. When the auto-register timeout expires and a new agent kicks in, and an instance ID in AWS has not come to GoCD, then terminate. That should reduce the likelihood of terminating instances that should be left running.

**Comment by brettcave** [comment]
sorry, i was mistaken about a cancelled pipeline continuing to launch agents. Was running a 2nd gocd server. But continuously launching new agents still occurs on an active pipeline, and launched instances are never terminated if they don't register.

**Comment by brettcave** [comment]
@v-sky - this happens when the agent is provisioned in the last provided subnet (which always occurs if only 1 subnet is provided).

in Ec2Instance.create, why does `i < subnets.size()` have any relevance? i is incremented in the finally, so for every loop, and there's a good chance i will equal the number of subnets provided.

**Comment by brettcave** [comment]
This stops errors for a single-subnet profile: https://github.com/brettcave/GoCD-EC2-Elastic-Agent-Plugin/pull/1/commits/2ed0df323e28cfff20060082f3707006bd31a58f
Have not tested with > 1 subnet.

**Comment by v-sky** [comment]
Issue fixed in e24a9921290dba6ee123253e307a70d40872a188
